### ISSUE TITLE
Unable to scale by 0

### ISSUE THREAD

**Issue by Enchufa2** [Unidata/UDUNITS-2#117]
Scaling by 0 gives:

```c
ut_scale(0, unit)
#> ut_scale(): NULL factor argument
```

I think this should work. Or how are we supposed to get 0 units?

**Comment by semmerson** [comment]
Hi Iñaki Ucar,

There's no such thing as a zero unit. There is the dimensionless unit "1",
however.

**Comment by Enchufa2** [comment]
So there's no way to represent 0 m? Or, fot that matter, 0 ºC? But converting 273.15 K to ºC works, so... I'm a bit confused here. :)

**Comment by Enchufa2** [comment]
Correct. And parsing "0 ºC" does not work with the "NULL factor argument" above. But parsing "273.15 K" does work. I just don't understand why a factor equal to 0 is special here in any way. Probably because I don't know udunits' internals, but still, as a user of the API, this is very confusing. And the documentation says nothing about this, just that it "_returns a unit equivalent to another unit scaled by a numeric factor_". 

For context, [here](https://github.com/r-quantities/units/blob/f7088ae66c82507c7bddd856c2f27b3992d9b1e7/src/udunits.cpp#L82-L111) we have a function where two vectors of numbers, with two different units, are compared. Particularly, it feels natural to just do [this](https://github.com/r-quantities/units/blob/f7088ae66c82507c7bddd856c2f27b3992d9b1e7/src/udunits.cpp#L97-L104):

```c++
  ut_unit *ux = ut_parse(sys, ut_trim(xn[0], enc), enc);
  ut_unit *uy = ut_parse(sys, ut_trim(yn[0], enc), enc);

  for (int i=0, j=0; i < x.size(); i++, j++) {
    if (j == y.size())
      j = 0;
    out[i] = ut_compare(ut_scale(x[i], ux), ut_scale(y[j], uy));
  }
```

I.e., units `xn` and `yn`, for vectors `x` and `y` respectively, are parsed _once_, and then scaled and compared by every element in those vectors (with index recycling). Let's forget that I'm not freeing the output of `ut_scale` and I should, but otherwise, this works for any factor except for 0.

**Comment by Enchufa2** [comment]
Also for reference, I'm currently applying [this fix](https://github.com/r-quantities/units/commit/66b92181813e8df3d02c7c0b535ea35cf67c53b8#diff-c6feb88adee4971fc91670e49d4f554c0f90a4dc284d028e334421b3f7cc65f2), because, admittedly, if one unit is equal to 0, then you don't need the unit to compare them. But this works if units are not compatible, when it shouldn't. And in general it's unintuitive, and having to apply this exception doesn't feel right.

**Comment by Enchufa2** [comment]
The issue here may be that I don't think that `ut_compare` works as I expect...

```r
units:::ud_compare(0.5, c(-0.1, 1, 2)*39.3701, "m", "inch")
#> [1] -1 -1 -1
```

This is weird to me. It means basically that `ut_compare` is not meant to compare quantities, but just "unit objects", whatever it means, and then I don't understand what's the ordering. I thought that this was for comparing quantities when the use of `ut_compare` was suggested in [this issue](https://github.com/Unidata/UDUNITS-2/issues/113#issuecomment-1360568168). So I'm a bit lost now.

**Comment by larsbarring** [comment]
UDUNITS is a *units* package , not a *quantity* package. You can have a quantity that is "42 m", where "42" is the amount and "m" is the unit. In general terms this is often expressed as `Z = {Z} · [Z]` where `Z` is the quantity, `{Z}` is the amount, and `[Z]` is the unit. Now, what would a *unit* [0 m] mean? As [already explained](https://github.com/Unidata/UDUNITS-2/issues/117#issuecomment-1520394953):
> There's no such thing as a zero unit. There is the dimensionless unit "1", however.

As a *units* package UDUNITS does not compare quantities (or carry out any other operations on quantities). As was already explained in [this #113 comment](https://github.com/Unidata/UDUNITS-2/issues/113#issuecomment-1362062496)
> ut_compare() is for putting all units in a global order so that, for example, they can be inserted into a binary tree.

Suggest close as answered.

**Comment by Enchufa2** [comment]
Forget about the quantity vs. unit, magnitude and whatnot. The question is, what is the difference between parsing "0 ºC" and parsing "1 ºC" then scaling it by 0?

**Comment by larsbarring** [comment]
I think most of the confusion here comes from a not-so-informative error message and a unit-vs-amount distinction that's easy to trip over. Let me try to untangle it.

**The error message.** `ut_scale(0, unit)` reports `"NULL factor argument"`, but of course `0.0` isn't a null pointer. The real reason for the rejection is that a unit in UDUNITS has to be *invertible*: every unit must support converting values both into and out of it. A unit scaled by zero would map every value to the same point, so there's no way back — no converter could ever be built from it. That's why a zero scale is refused. The message should say something like "a unit cannot have a zero scale factor"; sorry it sent you down the wrong path.

**`ut_scale` scales the unit, not a value.** `ut_scale(60, second)` doesn't mean "60 seconds" (a quantity); it produces a *new unit* sixty times the size of a second — i.e. a unit equivalent to the minute. So when you write `ut_scale(x[i], ux)`, you aren't expressing the value `x[i]` in `ux`; you're manufacturing a brand-new unit (and asking for one with a zero scale is what trips the error). That's not what you want for comparing measurements.

**Your direct question (parse `"0 °C"` vs. parse `"1 °C"` then scale by 0):** there's actually no difference — they're the *same* operation. The grammar turns `<number> <unit>` into a scaling, so `"0 °C"` is parsed as `ut_scale(0, °C)`, which is exactly what you did by hand. Both are refused for the same invertibility reason. So the library is being consistent, not arbitrary; ºC isn't special here.

**What `ut_compare` is for (the non-technical version).** Think of it like alphabetising entries in an index. To store units in a fast lookup table, the library needs a fixed, consistent way to put *any* two units in some order — including units that have nothing to do with each other. That order is deliberate and stable, but it carries no physical meaning. Just as alphabetical order puts "metre" before "second" without implying a metre is somehow smaller, `ut_compare` says nothing about magnitude. That's why "0.5 m" can sort above "1 m", and why it even returns an order for metres vs. inches: it has to, to be a usable sort key. It was never meant to compare measurements.

**How to actually compare quantities.** UDUNITS converts units; comparing the *numbers* is the caller's job. Check compatibility once with `ut_are_convertible`, build one converter, convert into a common unit, then compare plain numbers:

```c
ut_unit *ux = ut_parse(sys, "m", enc);
ut_unit *uy = ut_parse(sys, "inch", enc);

if (!ut_are_convertible(ux, uy)) {
    /* genuinely not comparable — handle as you wish */
}

cv_converter *uy_to_ux = ut_get_converter(uy, ux);

for (int i = 0, j = 0; i < n; i++, j = (j + 1 == m) ? 0 : j + 1) {
    double y_in_ux = cv_convert_double(uy_to_ux, y[j]);
    out[i] = (x[i] < y_in_ux) ? -1 : (x[i] > y_in_ux) ? 1 : 0;
}
cv_free(uy_to_ux);
```

`ut_are_convertible` is the explicit "are these two units comparable at all?" test — the check you were reaching for. It also removes the zero/one special cases entirely and handles offset units like °C correctly, since the converter applies the offset for you.

Hope that clears it up.

**Comment by Enchufa2** [comment]
Thank you, _this_ is very helpful. Now I think I understand it better.

About comparing quantities, the problem is that `ut_are_convertible` will say yes for e.g. "m" and "m", and we didn't want to waste any more cycles in these cases. So we currently use `ut_compare` just to tell whether two units are the same or not, and avoid such cases. See [here](https://github.com/r-quantities/units/blob/3854ba01d62953e9d10b5aabd064cd887fd0383d/src/udunits.cpp#L83). I hope this is fine. At least it is working fine. Which brings me to question the utility of `ut_compare` as part of the C API. IMHO if the sole purpose of this function is to sort units, with no meaning at all, then from the user perspective it is confusing, and it would be much more clear and useful to have a function `ut_is_same`. Or at least document `ut_compare` saying that you should use it for this, checking whether the output is 0 or not, and then if you want the technical details, this is what -1 and 1 means.

About values vs. factors, the confusion comes from the fact that when you parse e.g. "35 m" you get a unit, and when you format that unit, you get "35 m" back. More so, if you parse "-35 m" and format it, you get back "-35 m"! So I know now that this -35 is not a value... because you told me so. But from the user perspective, it is very much like the unit is holding a value, you see? And that's why getting an error for trying to parse "0 m" is so confusing. And I mentioned ºC because as you know temperature, conceptually, is kind of difficult and special in this regard.

So we stopped scaling units and comparing in this way, as you can see in the link above, but still I would suggest adding some explanation in the documentation. And I agree that the error should not mention a null pointer.

**Comment by Enchufa2** [comment]
BTW, feel free to close this, unless you are using it as a reminder for adding something to the docs.
